// minimatch/dist/commonjs/ast.js
"use strict";
// parse a single path portion
Object.defineProperty(exports, "__esModule", { value: true });
exports.AST = void 0;
const brace_expressions_js_1 = require("./brace-expressions.js");
const unescape_js_1 = require("./unescape.js");
const types = new Set(['!', '?', '+', '*', '@']);
const isExtglobType = (c) => types.has(c);
// Patterns that get prepended to bind to the start of either the
// entire string, or just a single path portion, to prevent dots
// and/or traversal patterns, when needed.
// Exts don't need the ^ or / bit, because the root binds that already.
const startNoTraversal = '(?!(?:^|/)\\.\\.?(?:$|/))';
const startNoDot = '(?!\\.)';
// characters that indicate a start of pattern needs the "no dots" bit,
// because a dot *might* be matched. ( is not in the list, because in
// the case of a child extglob, it will handle the prevention itself.
const addPatternStart = new Set(['[', '.']);
// cases where traversal is A-OK, no dot prevention needed
const justDots = new Set(['..', '.']);
const reSpecials = new Set('().*{}+?[]^$\\!');
const regExpEscape = (s) => s.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&');
// any single thing other than /
const qmark = '[^/]';
// * => any number of characters
const star = qmark + '*?';
// use + when we need to ensure that *something* matches, because the * is
// the only thing in the path portion.
const starNoEmpty = qmark + '+?';
// remove the \ chars that we added if we end up doing a nonmagic compare
// const deslash = (s: string) => s.replace(/\\(.)/g, '$1')
class AST {
    type;
    #root;
    #hasMagic;
    #uflag = false;
    #parts = [];
    #parent;
    #parentIndex;
    #negs;
    #filledNegs = false;
    #options;
    #toString;
    // set to true if it's an extglob with no children
    // (which really means one child of '')
    #emptyExt = false;
    constructor(type, parent, options = {}) {
        this.type = type;
        // extglobs are inherently magical
        if (type)
            this.#hasMagic = true;
        this.#parent = parent;
        this.#root = this.#parent ? this.#parent.#root : this;
        this.#options = this.#root === this ? options : this.#root.#options;
        this.#negs = this.#root === this ? [] : this.#root.#negs;
        if (type === '!' && !this.#root.#filledNegs)
            this.#negs.push(this);
        this.#parentIndex = this.#parent ? this.#parent.#parts.length : 0;
    }
    get hasMagic() {
        /* c8 ignore start */
        if (this.#hasMagic !== undefined)
            return this.#hasMagic;
        /* c8 ignore stop */
        for (const p of this.#parts) {
            if (typeof p === 'string')
                continue;
            if (p.type || p.hasMagic)
                return (this.#hasMagic = true);
        }
        // note: will be undefined until we generate the regexp src and find out
        return this.#hasMagic;
    }
    // reconstructs the pattern
    toString() {
        if (this.#toString !== undefined)
            return this.#toString;
        if (!this.type) {
            return (this.#toString = this.#parts.map(p => String(p)).join(''));
        }
        else {
            return (this.#toString =
                this.type + '(' + this.#parts.map(p => String(p)).join('|') + ')');
        }
    }
    #fillNegs() {
        /* c8 ignore start */
        if (this !== this.#root)
            throw new Error('should only call on root');
        if (this.#filledNegs)
            return this;
        /* c8 ignore stop */
        // call toString() once to fill this out
        this.toString();
        this.#filledNegs = true;
        let n;
        while ((n = this.#negs.pop())) {
            if (n.type !== '!')
                continue;
            // walk up the tree, appending everthing that comes AFTER parentIndex
            let p = n;
            let pp = p.#parent;
            while (pp) {
                for (let i = p.#parentIndex + 1; !pp.type && i < pp.#parts.length; i++) {
                    for (const part of n.#parts) {
                        /* c8 ignore start */
                        if (typeof part === 'string') {
                            throw new Error('string part in extglob AST??');
                        }
                        /* c8 ignore stop */
                        part.copyIn(pp.#parts[i]);
                    }
                }
                p = pp;
                pp = p.#parent;
            }
        }
        return this;
    }
    push(...parts) {
        for (const p of parts) {
            if (p === '')
                continue;
            /* c8 ignore start */
            if (typeof p !== 'string' && !(p instanceof AST && p.#parent === this)) {
                throw new Error('invalid part: ' + p);
            }
            /* c8 ignore stop */
            this.#parts.push(p);
        }
    }
    toJSON() {
        const ret = this.type === null
            ? this.#parts.slice().map(p => (typeof p === 'string' ? p : p.toJSON()))
            : [this.type, ...this.#parts.map(p => p.toJSON())];
        if (this.isStart() && !this.type)
            ret.unshift([]);
        if (this.isEnd() &&
            (this === this.#root ||
                (this.#root.#filledNegs && this.#parent?.type === '!'))) {
            ret.push({});
        }
        return ret;
    }
    isStart() {
        if (this.#root === this)
            return true;
        // if (this.type) return !!this.#parent?.isStart()
        if (!this.#parent?.isStart())
            return false;
        if (this.#parentIndex === 0)
            return true;
        // if everything AHEAD of this is a negation, then it's still the "start"
        const p = this.#parent;
        for (let i = 0; i < this.#parentIndex; i++) {
            const pp = p.#parts[i];
            if (!(pp instanceof AST && pp.type === '!')) {
                return false;
            }
        }
        return true;
    }
    isEnd() {
        if (this.#root === this)
            return true;
        if (this.#parent?.type === '!')
            return true;
        if (!this.#parent?.isEnd())
            return false;
        if (!this.type)
            return this.#parent?.isEnd();
        // if not root, it'll always have a parent
        /* c8 ignore start */
        const pl = this.#parent ? this.#parent.#parts.length : 0;
        /* c8 ignore stop */
        return this.#parentIndex === pl - 1;
    }
    copyIn(part) {
        if (typeof part === 'string')
            this.push(part);
        else
            this.push(part.clone(this));
    }
    clone(parent) {
        const c = new AST(this.type, parent);
        for (const p of this.#parts) {
            c.copyIn(p);
        }
        return c;
    }
    static #parseAST(str, ast, pos, opt) {
        let escaping = false;
        let inBrace = false;
        let braceStart = -1;
        let braceNeg = false;
        if (ast.type === null) {
            // outside of a extglob, append until we find a start
            let i = pos;
            let acc = '';
            while (i < str.length) {
                const c = str.charAt(i++);
                // still accumulate escapes at this point, but we do ignore
                // starts that are escaped
                if (escaping || c === '\\') {
                    escaping = !escaping;
                    acc += c;
                    continue;
                }
                if (inBrace) {
                    if (i === braceStart + 1) {
                        if (c === '^' || c === '!') {
                            braceNeg = true;
                        }
                    }
                    else if (c === ']' && !(i === braceStart + 2 && braceNeg)) {
                        inBrace = false;
                    }
                    acc += c;
                    continue;
                }
                else if (c === '[') {
                    inBrace = true;
                    braceStart = i;
                    braceNeg = false;
                    acc += c;
                    continue;
                }
                if (!opt.noext && isExtglobType(c) && str.charAt(i) === '(') {
                    ast.push(acc);
                    acc = '';
                    const ext = new AST(c, ast);
                    i = AST.#parseAST(str, ext, i, opt);
                    ast.push(ext);
                    continue;
                }
                acc += c;
            }
            ast.push(acc);
            return i;
        }
        // some kind of extglob, pos is at the (
        // find the next | or )
        let i = pos + 1;
        let part = new AST(null, ast);
        const parts = [];
        let acc = '';
        while (i < str.length) {
            const c = str.charAt(i++);
            // still accumulate escapes at this point, but we do ignore
            // starts that are escaped
            if (escaping || c === '\\') {
                escaping = !escaping;
                acc += c;
                continue;
            }
            if (inBrace) {
                if (i === braceStart + 1) {
                    if (c === '^' || c === '!') {
                        braceNeg = true;
                    }
                }
                else if (c === ']' && !(i === braceStart + 2 && braceNeg)) {
                    inBrace = false;
                }
                acc += c;
                continue;
            }
            else if (c === '[') {
                inBrace = true;
                braceStart = i;
                braceNeg = false;
                acc += c;
                continue;
            }
            if (isExtglobType(c) && str.charAt(i) === '(') {
                part.push(acc);
                acc = '';
                const ext = new AST(c, part);
                part.push(ext);
                i = AST.#parseAST(str, ext, i, opt);
                continue;
            }
            if (c === '|') {
                part.push(acc);
                acc = '';
                parts.push(part);
                part = new AST(null, ast);
                continue;
            }
            if (c === ')') {
                if (acc === '' && ast.#parts.length === 0) {
                    ast.#emptyExt = true;
                }
                part.push(acc);
                acc = '';
                ast.push(...parts, part);
                return i;
            }
            acc += c;
        }
        // unfinished extglob
        // if we got here, it was a malformed extglob! not an extglob, but
        // maybe something else in there.
        ast.type = null;
        ast.#hasMagic = undefined;
        ast.#parts = [str.substring(pos - 1)];
        return i;
    }
    static fromGlob(pattern, options = {}) {
        const ast = new AST(null, undefined, options);
        AST.#parseAST(pattern, ast, 0, options);
        return ast;
    }
    // returns the regular expression if there's magic, or the unescaped
    // string if not.
    toMMPattern() {
        // should only be called on root
        /* c8 ignore start */
        if (this !== this.#root)
            return this.#root.toMMPattern();
        /* c8 ignore stop */
        const glob = this.toString();
        const [re, body, hasMagic, uflag] = this.toRegExpSource();
        // if we're in nocase mode, and not nocaseMagicOnly, then we do
        // still need a regular expression if we have to case-insensitively
        // match capital/lowercase characters.
        const anyMagic = hasMagic ||
            this.#hasMagic ||
            (this.#options.nocase &&
                !this.#options.nocaseMagicOnly &&
                glob.toUpperCase() !== glob.toLowerCase());
        if (!anyMagic) {
            return body;
        }
        const flags = (this.#options.nocase ? 'i' : '') + (uflag ? 'u' : '');
        return Object.assign(new RegExp(`^${re}$`, flags), {
            _src: re,
            _glob: glob,
        });
    }
    get options() {
        return this.#options;
    }
    // returns the string match, the regexp source, whether there's magic
    // in the regexp (so a regular expression is required) and whether or
    // not the uflag is needed for the regular expression (for posix classes)
    // TODO: instead of injecting the start/end at this point, just return
    // the BODY of the regexp, along with the start/end portions suitable
    // for binding the start/end in either a joined full-path makeRe context
    // (where we bind to (^|/), or a standalone matchPart context (where
    // we bind to ^, and not /).  Otherwise slashes get duped!
    //
    // In part-matching mode, the start is:
    // - if not isStart: nothing
    // - if traversal possible, but not allowed: ^(?!\.\.?$)
    // - if dots allowed or not possible: ^
    // - if dots possible and not allowed: ^(?!\.)
    // end is:
    // - if not isEnd(): nothing
    // - else: $
    //
    // In full-path matching mode, we put the slash at the START of the
    // pattern, so start is:
    // - if first pattern: same as part-matching mode
    // - if not isStart(): nothing
    // - if traversal possible, but not allowed: /(?!\.\.?(?:$|/))
    // - if dots allowed or not possible: /
    // - if dots possible and not allowed: /(?!\.)
    // end is:
    // - if last pattern, same as part-matching mode
    // - else nothing
    //
    // Always put the (?:$|/) on negated tails, though, because that has to be
    // there to bind the end of the negated pattern portion, and it's easier to
    // just stick it in now rather than try to inject it later in the middle of
    // the pattern.
    //
    // We can just always return the same end, and leave it up to the caller
    // to know whether it's going to be used joined or in parts.
    // And, if the start is adjusted slightly, can do the same there:
    // - if not isStart: nothing
    // - if traversal possible, but not allowed: (?:/|^)(?!\.\.?$)
    // - if dots allowed or not possible: (?:/|^)
    // - if dots possible and not allowed: (?:/|^)(?!\.)
    //
    // But it's better to have a simpler binding without a conditional, for
    // performance, so probably better to return both start options.
    //
    // Then the caller just ignores the end if it's not the first pattern,
    // and the start always gets applied.
    //
    // But that's always going to be $ if it's the ending pattern, or nothing,
    // so the caller can just attach $ at the end of the pattern when building.
    //
    // So the todo is:
    // - better detect what kind of start is needed
    // - return both flavors of starting pattern
    // - attach $ at the end of the pattern when creating the actual RegExp
    //
    // Ah, but wait, no, that all only applies to the root when the first pattern
    // is not an extglob. If the first pattern IS an extglob, then we need all
    // that dot prevention biz to live in the extglob portions, because eg
    // +(*|.x*) can match .xy but not .yx.
    //
    // So, return the two flavors if it's #root and the first child is not an
    // AST, otherwise leave it to the child AST to handle it, and there,
    // use the (?:^|/) style of start binding.
    //
    // Even simplified further:
    // - Since the start for a join is eg /(?!\.) and the start for a part
    // is ^(?!\.), we can just prepend (?!\.) to the pattern (either root
    // or start or whatever) and prepend ^ or / at the Regexp construction.
    toRegExpSource(allowDot) {
        const dot = allowDot ?? !!this.#options.dot;
        if (this.#root === this)
            this.#fillNegs();
        if (!this.type) {
            const noEmpty = this.isStart() && this.isEnd();
            const src = this.#parts
                .map(p => {
                const [re, _, hasMagic, uflag] = typeof p === 'string'
                    ? AST.#parseGlob(p, this.#hasMagic, noEmpty)
                    : p.toRegExpSource(allowDot);
                this.#hasMagic = this.#hasMagic || hasMagic;
                this.#uflag = this.#uflag || uflag;
                return re;
            })
                .join('');
            let start = '';
            if (this.isStart()) {
                if (typeof this.#parts[0] === 'string') {
                    // this is the string that will match the start of the pattern,
                    // so we need to prot?ÿlJ-ÿ]D0ÿF=7ÿXK@ÿvn_ÿ„Šÿ— ÿ«²´ÿ¶»¿ÿ¹¾Âÿ¼ÀÄÿ¿ÃÆÿÂÄÈÿÅÈÉÿÈÊËÿËÌÌÿÎÎÎÿÑÑĞÿÔÓÒÿ×ÕÔÿÚØÕÿİÚØÿàİÚÿãßÜÿæâŞÿéäàÿëæáÿìçãÿíéåÿîêçÿïìéÿñíëÿòïìÿîìéÿØÙÖÿ®µ°ÿš¢›ÿ³´°ÿ¦£¡ÿ¥¢¡ÿëêéÿûûûÿóóòÿééèÿ¬¬©õkkf¤			Q   <   !   
                                                                                                                                                             XXUH†~uà¶©ÿs]ÿkH+ÿeH1ÿJ?7ÿOB8ÿum_ÿˆ†ÿ›££ÿ¬³¸ÿ²¹¿ÿ´»Àÿ¶¼Áÿº¿Ãÿ¼ÁÄÿ¿ÂÆÿÂÅÈÿÄÇÉÿÈÊËÿÊÌÍÿÍÍÎÿÏĞÏÿÒÒÑÿÕÔÓÿØ×ÕÿÛÙ×ÿŞÜÙÿáŞÛÿäáİÿçãßÿêåáÿëèäÿíéæÿïëèÿğìêÿòïìÿóğîÿôòğÿöôòÿ÷öôÿììêÿÀÆÂÿ ©£ÿ´¸´ÿ§¦¤ÿ®­¬ÿâááÿââáÿããâÿÌÌÊş„„Ó553i   D   *                                                                                                                                                     cc^i”Š€ú¶¥—ÿoL0ÿjH,ÿ]L>ÿPHBÿoeUÿ†‡ÿ–œÿª²¹ÿ­µ¼ÿ¯·¾ÿ²¹¿ÿ´»Àÿ·½Âÿ¹¿Ãÿ¼ÀÄÿ¾ÂÆÿ¾ÂÅÿ¼¿Áÿ·¸¹ÿ²³²ÿ¯¯­ÿ­¬«ÿ¯®¬ÿ®­«ÿ°®¬ÿ±°®ÿµ³±ÿ´²°ÿ¸¶³ÿ½»·ÿÆÃÀÿÕÒÏÿáŞÛÿëéæÿïìéÿñîìÿóñîÿõóñÿ÷õóÿø÷õÿúøøÿûúúÿùùøÿÀÆÂÿ§¡ÿµ·´ÿ”’ÿ»º¸ÿÒÒĞÿĞĞÎÿÙÙ×ÿ™™•ôUUS†   J   1                                                                                                                                             mjc¨’û…rÿjG*ÿiL4ÿ^UOÿgYLÿ‚wÿ–”ÿ£­´ÿ§²ºÿª´»ÿ­µ½ÿ¯·¾ÿ²¹¿ÿ´»Àÿµ¼Áÿ³¹½ÿ¦©ªÿ¢££ÿ¦§§ÿ«¯®ÿ°µµÿ´ººÿ·½»ÿ¹¿½ÿ»Á¿ÿ½ÃÀÿ¿ÅÂÿÂÇÃÿÄÊÅÿÇËÇÿÉÌÈÿÉÌÈÿÊÌÈÿÃÄÀÿº»·ÿ¯®¬ÿ·µ²ÿÔÓÑÿíëéÿóòğÿöõôÿù÷öÿûúùÿüüûÿşşıÿşşşÿíïíÿ°·³ÿ¦ ÿ¢¢ ÿ¡ ÿÁÁ¿ÿ¿¿½ÿÑÑÏÿ¨¨¥øggd N   5                                                                                                                                        uogª´§œş‡kSÿiF)ÿlVFÿg]Vÿvpcÿ‡‡ÿš¢§ÿ£¯¸ÿ¥°¹ÿ§²»ÿª´¼ÿ¬¶½ÿ®·¾ÿ­µ»ÿ•šÿŒÿ—›œÿ¦­®ÿ­µ·ÿ­¶·ÿ®·¸ÿ°¹¸ÿ²»ºÿµ½¼ÿ·¾½ÿ¹À¾ÿ¼ÂÀÿ¾ÄÁÿÀÆÃÿÂÈÄÿÃÉÅÿÄËÅÿÇÌÈÿÉÎÉÿËÏËÿÍÑÍÿĞÔĞÿÎÒÎÿÁÃÀÿ²²°ÿ´³±ÿÛÚÙÿõõôÿúùùÿüüüÿşşşÿüüüÿúúùÿ÷÷÷ÿÕÙÖÿ™£ÿ£§£ÿ¦¥£ÿµ´±ÿ±±®ÿÁÂ¾ÿ³³¯ıtto´R   6                                                                                                                             ***xsi¼·ªŸÿxY>ÿjG*ÿqaUÿpe\ÿ~sÿ“ÿ¡«³ÿ¢®¸ÿ£¯¸ÿ¦±ºÿ§²»ÿ©´¼ÿ¥®µÿ…‰ÿy{~ÿ—Ÿ¢ÿ§²µÿ§²µÿ©³µÿ«µ·ÿ¯¸¹ÿ´¼¼ÿ·¾¿ÿ»ÂÂÿÀÅÆÿÃÈÈÿÆÊÉÿÈÍÌÿÌÏÎÿÎÑÏÿÎÑÏÿÎÒÏÿÏÒÏÿÏÓÏÿÏÓÏÿÏÓÎÿÎÒÎÿÍÒÎÿÎÓÏÿÏÕÑÿÒØÔÿÔØÕÿ·¹¶ÿ¨¨¦ÿÔÔÓÿ÷ø÷ÿûüûÿúúúÿøøøÿõõõÿóóòÿëëêÿ¨°«ÿš¡œÿ¯®¬ÿ±±®ÿ¦¦¢ÿ³³¯ÿµµ±ÿuuq½R   7                                                                                                                     (((zsjº¸©ÿzZ@ÿjH,ÿzogÿuj]ÿ‚‡}ÿ˜œ›ÿ¢­¶ÿ£®¸ÿ£®¸ÿ£¯¹ÿ¥°ºÿ¤¯¸ÿy~„ÿoswÿ•£ÿ£¯³ÿ¤°³ÿ§²µÿ«µ¸ÿ¯¸¼ÿ´½¿ÿºÁÄÿ¿ÄÇÿÂÈÊÿÆËÍÿËÎĞÿÎĞÒÿÑÓÔÿÔÕÖÿ×ØØÿÚÚÚÿÜÜÛÿİİÜÿßßİÿßàİÿààİÿŞßÜÿÜİÚÿØÛ×ÿÔÙÕÿÒ×ÔÿÑ×ÓÿÑØÔÿÔÚ×ÿÕÛØÿÁÄÂÿ§§¥ÿÚÛÚÿôôôÿõõõÿóôóÿñññÿïïîÿììëÿ¿ÄÁÿ’›•ÿµ´²ÿ··´ÿœœ˜ÿ©©¥ÿ³³¯şuuq»R   5                                                                                                                zum­¸«ŸÿvV:ÿkI-ÿvnÿzpcÿ…Šÿ£§©ÿ¯¹Àÿ¤¯¸ÿ£¯¸ÿ£¯¹ÿ£¯¹ÿ‡˜ÿaejÿŠ”šÿŸ­²ÿ¡®³ÿ¤±¶ÿªµºÿ°¹¾ÿ´¼Áÿ¹ÀÅÿ¼ÃÈÿÀÆËÿÃÉÍÿÇÌÏÿÊÎÑÿÍÑÓÿĞÓÕÿÔÖ×ÿÖØÙÿÚÚÛÿİİİÿßßßÿãââÿæåãÿèçåÿêéçÿìëèÿìëèÿêêçÿæçäÿàãàÿÚßÜÿ×ÜÙÿÔÛØÿÔÛÙÿÖİÛÿÕÛÙÿ·º¹ÿ®®¬ÿâäãÿïğïÿïïïÿîîíÿììëÿêêéÿÍĞÎÿ™“ÿ´µ²ÿ¾¾»ÿ““ÿ¢¢ÿ®¯«şsso·			O   2                                                                                                      
tqj›±£—ÿuU9ÿjH+ÿ‡|tÿ|qcÿ…‹‚ÿ¥©ªÿ²»Âÿ°ºÁÿ¤¯¸ÿ¤¯¸ÿŸ«´ÿsyÿw„ÿœª±ÿ¬²ÿ¢¯µÿ¨³ºÿ­·¾ÿ±ºÁÿµ½Äÿ¸ÀÇÿ»ÂÉÿ¾ÄËÿÁÈÌÿÄÊÎÿÇÌĞÿËÏÒÿÎÑÔÿÑÓÖÿÔÖØÿ×ÙÚÿÚÛÜÿİŞŞÿààáÿãããÿææåÿêèèÿìëêÿğîìÿòğîÿôòğÿóòğÿññïÿíîìÿæêèÿßäâÿÚàŞÿÕÜÚÿÒÙ×ÿÒÙ×ÿÍÓĞÿ¤¤¢ÿËËÊÿéêéÿëëëÿêêéÿééçÿææåÿÖØÖÿš”ÿ·¸µÿÂÁ¾ÿ‹‹†ÿœœ—ÿ££Ÿÿkkg¨   K   ,                                                                                                 nlej®¡—üƒfMÿjH+ÿ…~ÿuhÿ†Š‚ÿ©®²ÿ³»Âÿ²»Âÿ²»Âÿ¥°¹ÿ˜¢ªÿoswÿ›¢ÿ›ª²ÿ¬´ÿ£°¸ÿ¨´¼ÿ¬·¿ÿ°ºÂÿ³¼Äÿ¶¾Æÿ¹ÀÈÿ»ÃÊÿ¾ÅËÿÁÈÍÿÄÊÏÿÇÍÑÿÊÏÓÿÍÒÕÿĞÔ×ÿÓ×Ùÿ×ÙÛÿÚÛİÿİŞßÿàáâÿããäÿæææÿêéèÿìëëÿğîíÿóñğÿöôòÿ÷öôÿø÷öÿøø÷ÿö÷öÿñóòÿéíìÿŞäâÿÕÜÙÿĞ×ÔÿÍÕÒÿÎÕÒÿ»¾¼ÿ¼½»ÿãåäÿèèçÿççæÿååäÿääâÿÓÕÓÿ˜’ÿ¼½ºÿ¿¾»ÿ‚‚}ÿŸŸšÿ™™”ú[[WŒ   F   #                                                                                          ggc@›…ù‘waÿiF(ÿwÿƒylÿ„‡}ÿ¦©«ÿ³»Âÿ³»Âÿ³¼Âÿ²»Âÿ•£ÿ~„ˆÿ›¨®ÿœ«²ÿŸ­¶ÿ¤±ºÿ¨´½ÿ«¶¿ÿ®¸Áÿ°ºÃÿ³½Åÿ¶¾Çÿ¸ÁÉÿ»ÃÊÿ¾ÅÌÿÁÈÎÿÄËĞÿÇÍÒÿÊÏÔÿÍÒÖÿĞÔØÿÓ×ÚÿÖÚÜÿÙÜŞÿÜßàÿàáâÿãäåÿææçÿéééÿìììÿğïîÿóòñÿöôóÿù÷öÿúùøÿûûúÿüüüÿüıüÿ÷ø÷ÿîğïÿâæäÿÖÜÚÿÎÖÓÿÊÒÏÿÊÒÏÿÆËÈÿÀÁ¿ÿİŞİÿäåãÿääãÿããáÿááàÿÑÓÑÿ—ÿÂÃÀÿ°°¬ÿ{{uÿ¤¤Ÿÿ‰Š…ñ==;p   ?                                                                                      ???‰€wŞ®›‹ÿhD&ÿ‡scÿŒ‚wÿ‚†|ÿ¡¢¡ÿ´¼Âÿ´¼Âÿ´¼Âÿ³¼Âÿ£¨­ÿ•™ÿ«²ÿ¬´ÿ¡®·ÿ¤±»ÿ¦³½ÿ¨µ¿ÿ«·Àÿ­¹Âÿ°ºÃÿ³½Åÿµ¾Èÿ¸ÁÉÿ»ÃËÿ½ÆÍÿÁÉÏÿÄËĞÿÇÍÒÿÊĞÔÿÍÒÖÿĞÔÙÿÓ×ÚÿÖÚİÿÙÜßÿÜßáÿßâãÿâäæÿæçèÿéêêÿììíÿïïïÿóòòÿöõôÿùø÷ÿûúùÿıüüÿşşşÿıııÿúúúÿöööÿïğïÿäçæÿ×ÜÛÿÎÔÑÿÈĞÍÿÇÎËÿÉÏÌÿÉÊÈÿİİÜÿââàÿáâàÿààßÿßßİÿÌÎÌÿŠ“ÿÑÒÏÿ——’ÿvvpÿ­­¨ÿ€€{ßW   4                                                                                 ysk¤´¤—ÿkH+ÿz_Hÿš’‹ÿ„zÿ˜—“ÿµ¼Âÿµ¼Âÿ´¼Âÿ´¼Âÿ¤¨­ÿ¡§«ÿ¨³ºÿ¡­µÿ£°¹ÿ¥±»ÿ¦²½ÿ¦³¾ÿ¨µÀÿ«·Áÿ­¹Ãÿ°»Äÿ³½Æÿµ¿Èÿ¸ÁÊÿ»ÃÌÿ½ÆÍÿÀÉÏÿÃËÑÿÇÍÓÿÊĞÕÿÍÓ×ÿĞÕÙÿÓØÛÿÖÚŞÿÙİàÿÜßâÿßâäÿâåçÿåçéÿèêìÿìíîÿïğğÿòóóÿõõõÿùùøÿüûûÿşııÿşşşÿüüüÿùùùÿöööÿóóòÿîïîÿåçåÿØÜÚÿÍÓĞÿÆÎËÿÄÌÉÿÈÎËÿÕÖÔÿİŞÜÿßßŞÿßßŞÿŞŞÜÿÜÜÛÿ½Á¾ÿ•ÿàßŞÿ€€{ÿ{{uÿ¢£ÿook²I   &                                                                          olfP®¢˜ü€bIÿpP5ÿ¦˜ÿ~rÿŒ„ÿµ»Àÿ¶¼Âÿµ¼Âÿµ¼Âÿª¯³ÿ¦«°ÿ±ºÀÿ®¸¿ÿ¥±ºÿ¦²»ÿ¦³½ÿ¦³¾ÿ§´¿ÿ¨µÀÿ«·Áÿ­¹Ãÿ°»Äÿ²½Çÿµ¿Éÿ¸ÁËÿ»ÄÌÿ½ÆÎÿÀÉĞÿÃÌÒÿÆÎÔÿÊĞÖÿÍÓØÿÏÕÚÿÓØÜÿÖÛŞÿÙİáÿÜàãÿßãåÿâåèÿåèêÿèëíÿëíïÿïğñÿòóôÿõööÿøùùÿüüüÿşşşÿıııÿúúúÿ÷÷÷ÿõõôÿòòòÿğğïÿììëÿäæäÿØÜÚÿÌÒÏÿÃËÈÿÂÊÆÿÇÍÊÿßàßÿÛİÛÿİİÛÿİİÛÿÜÜÚÿÚÚÙÿ¬²®ÿ—œ—ÿÒÒÏÿnngÿ‡ÿŠùHHFx   >                                                                      KKK‹‚xã¢‹yÿiE'ÿ†ÿˆ€tÿ……{ÿ½ÂÆÿºÀÆÿ¶¼Âÿµ¼Âÿ°´¹ÿ©­±ÿ´¼Ãÿ³¼Ãÿ²¼Ãÿ¨²»ÿ§³¼ÿ§³½ÿ§´¾ÿ§´¿ÿ¨µÀÿ«·Âÿ­¹Ãÿ¯»Åÿ²½Èÿµ¿Éÿ·ÂËÿºÄÍÿ½ÇÏÿÀÉÑÿÃÌÓÿÆÎÕÿÉÑ×ÿÌÓÙÿÏÖÛÿÒØİÿÕÛßÿØİâÿÛàäÿßãæÿâåèÿåèëÿèëîÿëîğÿîñóÿòôõÿõöøÿøùúÿûüıÿüııÿûûûÿøøøÿööõÿóóóÿññğÿïïîÿìììÿééèÿãäâÿ×ÚØÿÊĞÍÿÂÊÆÿÀÈÄÿÆÌÉÿäåãÿÙÚØÿÛÛÙÿÛÛÙÿÚÚØÿÖÖÔÿ•Ÿ™ÿ¹¼¸ÿ§§£ÿggaÿ˜ÿ||vÜS   .   
                                                               vri{¶¨œÿmJ-ÿƒiTÿ —ÿ€ƒwÿ¨§¦ÿÈÍÑÿÂÇÌÿ·¼Âÿ³¸¾ÿ«®²ÿµ¼Ãÿµ½Äÿµ½Äÿµ¾Åÿªµ¾ÿ¨´½ÿ¨´¾ÿ¨µ¿ÿ¨µÀÿ¨µÁÿª·Ãÿ¬¹Äÿ¯»Æÿ²¾Èÿ´ÀÊÿ·ÂÌÿºÄÎÿ½ÇĞÿ¿ÊÑÿÂÌÔÿÆÎÕÿÉÑ×ÿÌÔÚÿÏÖÜÿÒÙŞÿÕÛàÿØŞãÿÛáåÿŞãçÿáæêÿåéìÿèìïÿëîñÿîñóÿñôöÿõ÷ùÿøúûÿúüıÿùúûÿøùùÿ÷÷öÿôôôÿòòñÿğğïÿíííÿëëêÿééèÿææåÿàáàÿÕØÖÿÉÏÌÿÀÈÄÿ¾ÆÂÿÇÍÉÿäåäÿ×ØÖÿÙÚØÿÙÙ×ÿØØÖÿÊÌÊÿˆ’‹ÿÖÖÕÿssmÿsslÿ™š”ş^^]”   B                                                              ddd˜„ñŒqZÿjH*ÿ­¤ÿ€sÿ’…ÿÇËÏÿÉÎÑÿÈÍÑÿ¸½Âÿ®°³ÿ¶¼Âÿ¶½Äÿ¶¾Åÿ¶¾Åÿ¶¾Çÿ¯¹Âÿ©´¾ÿ©µ¿ÿ©µ¿ÿ©¶Áÿ©¶Âÿª¸Ãÿ¬¹Åÿ¯¼Çÿ²¾Éÿ´ÀËÿ·ÂÌÿºÅÎÿ¼ÈĞÿ¿ÊÒÿÂÌÔÿÅÏÖÿÉÑÙÿÌÔÚÿÏÖİÿÒÙßÿÕÜáÿØŞäÿÛáæÿŞäèÿáæêÿäéíÿçìğÿêïòÿîòõÿñõ÷ÿô÷úÿ÷úüÿ÷ùûÿöøùÿõööÿôôôÿòòòÿğğğÿîîíÿììëÿêêéÿèèçÿææåÿääâÿŞßİÿÓÖÔÿÆÌÉÿ¾ÅÂÿ½ÄÀÿÍÒĞÿãäãÿÖ×ÕÿØØÖÿ××ÖÿÖÖÕÿ±¶²ÿ”š”ÿºº·ÿaaZÿ‰ÿ|îZ   0                                           
